seccomp/umount.go

```diff
@@ -113,19 +120,237 @@ func (u *umountSyscallInfo) process() (*sysResponse, error) {
+	if err != nil {
+		return false, u.tracer.createContinueResponse(u.reqId)
+	}
+	initProcMountNs, err := u.cntr.InitProc().MountNsInode()
+	if err != nil {
+		return false, u.tracer.createContinueResponse(u.reqId)
+	}
+
+	// Obtain the sys-container's root-path inode.
+	syscntrRootInode := u.cntr.InitProc().RootInode()
+
+	// If process' mount-ns matches the sys-container's one, then we can simply
+	// rely on the target's mountID to discern an immutable target from a
+	// regular one. Otherwise, we cannot rely on the mountID field, as the values
+	// allocated by kernel for these very mountpoints will differ in other mount
+	// namespaces.
+	if processMountNs == initProcMountNs {
+
+		if u.processInfo.Root() == "/" {
+			processRootInode := u.processInfo.RootInode()
+
+			// Scenario 1): no-unshare(mnt) & no-pivot() & no-chroot()
+			if processRootInode == syscntrRootInode {
+				if ok := u.cntr.IsImmutableMountID(info.MountID); ok == true {
+					logrus.Debugf("Rejected unmount operation on %s target (scenario 1)",
+						u.Target)
+					return false, u.tracer.createErrorResponse(u.reqId, syscall.EPERM)
+				}
+				return true, nil
+			}
+
+			// Scenario 2: no-unshare(mnt) & pivot() & no-chroot()
+			if processRootInode != syscntrRootInode {
+				if ok := u.cntr.IsImmutableMountID(info.MountID); ok == true {
+					logrus.Debugf("Rejected unmount operation on %s target (scenario 2)",
+						u.Target)
+					return false, u.tracer.createErrorResponse(u.reqId, syscall.EPERM)
+				}
+				return true, nil
+			}
+
+			return true, nil
+		}
+
+		if u.processInfo.Root() != "/" {
+			// We are dealing with a chroot'ed process, so obtain the inode of "/"
+			// as seen within the process' namespaces, and *not* the one associated
+			// to the process' root-path.
+			processRootInode, err := mip.ExtractInode("/")
+			if err != nil {
+				return false, u.tracer.createErrorResponse(u.reqId, syscall.EINVAL)
+			}
+
+			// Scenario 3: no-unshare(mnt) & no-pivot() & chroot()
+			if processRootInode == syscntrRootInode {
+				if ok := u.cntr.IsImmutableMountID(info.MountID); ok == true {
+					logrus.Debugf("Rejected unmount operation on %s target (scenario 3)",
+						u.Target)
+					return false, u.tracer.createErrorResponse(u.reqId, syscall.EPERM)
+				}
+				return true, nil
+			}
+
+			// Scenario 4: no-unshare(mnt) & pivot() & chroot()
+			if processRootInode != syscntrRootInode {
+				if ok := u.cntr.IsImmutableMountID(info.MountID); ok == true {
+					logrus.Debugf("Rejected unmount operation on %s target (scenario 4)",
+						u.Target)
+					return false, u.tracer.createErrorResponse(u.reqId, syscall.EPERM)
+				}
+				return true, nil
+			}
+
+			return true, nil
+		}
+
+	} else {
+
+		if u.processInfo.Root() == "/" {
+			processRootInode := u.processInfo.RootInode()
+
+			// Scenario 5): unshare(mnt) & no-pivot() & no-chroot()
+			if processRootInode == syscntrRootInode {
+				if mip.IsOverlapMount(info) {
```

Comment:
This assumes that an immutable mount can't possibly be an overlap mount (i.e., two or more mounts stacked over the same mountpoint). That's a risky assumption. Is there a more robust way of knowing if the overlap mount is in fact immutable or not?